[TASK] V11 compatibility

calien666 opened the pull request:
Update for TYPO3 version 11.
Fixes #5

andreaswolf committed:
[TASK] Add setup script for v11, remove v8/v9

andreaswolf committed:
[TASK] Make compatible with v11, remove v8/v9

andreaswolf committed:
[TASK] Use PHP 8.1 in ddev

andreaswolf committed:
[TASK] Use PhpStan 1.8+

andreaswolf committed:
[RECTOR] Apply Typo3LevelSetList::UP_TO_TYPO3_9

andreaswolf committed:
[TASK] Add Rector

andreaswolf committed:
[RECTOR] Apply Typo3LevelSetList::UP_TO_TYPO3_10

andreaswolf committed:
[RECTOR] Apply Typo3LevelSetList::UP_TO_TYPO3_11

andreaswolf committed:
[TASK] Adjust return type in BasePollObjectConverterV9

Riiiad committed:
[TASK] Update project structure

Riiiad committed:
[TASK] Add configuration for DI

Riiiad committed:
[TASK] Move Extbase mapping

Riiiad committed:
[TASK] Fix integer value bug

Riiiad committed:
[TASK] Update development setup

Riiiad committed:
[TASK] Switch to composer installer 4 and update rector path in ddev

Riiiad committed:
[TASK] Remove deleted options for cookies

https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/11.0/Feature-87301-SecureCookiesEnabledByDefault.html

Riiiad committed:
[TASK] Fix undefined arrays after 8.1 upgrade

Riiiad committed:
[TASK] Configure EXT:NumberedPagination for poll list

Riiiad committed:
[TASK] Create Events for Poll Controller

Riiiad committed:
[TASK] Fix empty poll options bug

Riiiad committed:
[TASK] Refactor code to implement PSR-14 Events instead of Signals

Riiiad committed:
[TASK] Fix types and Undefined array keys

Riiiad committed:
[TASK] Correct Pagination markup and styling

Riiiad committed:
[TASK] Remove Slots

Riiiad committed:
[TASK] Replace slot functionality with Listeners

Riiiad committed:
[TASK] Add event dispatcher to PollRepository and create PermissionCheckEvent class

Riiiad committed:
[TASK] Use better naming for Listener identifiers

Riiiad committed:
[TASK] Reorder constructor parameters and inject PersistenceManagerInterface

Riiiad committed:
[TASK] Add FindPollsEvent to dispatch event in PollRepository

Riiiad committed:
[TASK] Fix Pagination styling

Riiiad committed:
[TASK] Fix CGL Violations

Riiiad committed:
[TASK] Dont export testing and setup files

Riiiad committed:
[TASK] Fix redirect after poll is created

Riiiad committed:
[TASK] Remove strict types to make the events extendable

Riiiad committed:
[TASK] Add missing property

Riiiad committed:
[TASK] Enable slugs in BE for Debugging

Riiiad committed:
[TASK] Override pagination markup

Riiiad committed:
[TASK] Move admin check method to Service class

Riiiad committed:
[TASK] Add correct constrain for finished polls and list view update

Riiiad committed:
[TASK] Show correct list view with and without paginated items

Riiiad committed:
[TASK] Add basic bootstrap pagination and remove old markup

calien666 committed:
[TASK] Add use instead of full qualified namespaces

calien666 committed:
[TASK] Move inject methods to the top

To have a quick overview, the inject methods for repositories where
moved to the top of the class. This allows a faster view of how the
dependencies are injected

calien666 committed:
[TASK] clean-up ddev config

* Remove phpmyadmin dependency
* Remove auto-generated config files

calien666 committed:
[TASK] Update git config

calien666 committed:
[TASK] Refactor view var assignment

calien666 committed:
[TASK] Finalize events and code-style-fix

calien666 committed:
[TASK] Finalize events and code-style-fix

calien666 commented on Classes/Extbase/TypeConverter/BasePollObjectConverterV9.php:
This file could b removed, as the new version supports TYPO3 >= 11

calien666 committed:
[BUGFIX] Reset poll slug default to empty string

calien666 commented on Resources/Private/Partials/Poll/List/Table.html:
```diff
@@ -16,16 +15,18 @@
         <tbody>
-        <f:if condition="{polls -> f:count()} > 0">
+
+        <f:if condition="{pagination.paginator.paginatedItems -> f:count()} > 0">
```
IMHO a poll list without activated pagination will give an empty table here. Either add a pagination helper in Controller even if no pagination is set-up or adjust the template here to work with paginated and non-paginated lists.

PKuhlmay replied:
the pagination is now hidden, when there are only one page to be displayed.

calien666 commented on rector.php:
Are you sure this file is required inside the extension ass the execution of rector is done only once?

PKuhlmay replied:
IMO the execution of rector should be done via actions everytime something is pushed and not only once.
Reason is, that we could fetch issues which would be changed by rector automatically.

calien666 reviewed (commented):
@Riiiad Take a look here.

Commit:
[TASK] Add makefile for rendering documentation locally

Commit:
[TASK] Set version of extension to 0.10 and update typo3 dependencies

Commit:
[DOCS] Update and fix documentation

Commit:
[TASK] Update Readme - remove old ddev commands and set correct image sizes

Commit:
[TASK] Fix rector config

PKuhlmay commented on rector.php:
```diff
@@ -0,0 +1,59 @@
+    $rectorConfig->ruleWithConfiguration(ExtEmConfRector::class, [
+        ExtEmConfRector::ADDITIONAL_VALUES_TO_BE_REMOVED => [],
+    ]);
+    $rectorConfig->ruleWithConfiguration(ExtensionComposerRector::class, [
```
Done

PKuhlmay commented on rector.php:
```diff
@@ -0,0 +1,59 @@
+use Rector\Config\RectorConfig;
+use Rector\Core\ValueObject\PhpVersion;
+use Ssch\TYPO3Rector\Configuration\Typo3Option;
+use Ssch\TYPO3Rector\FileProcessor\Composer\Rector\ExtensionComposerRector;
```
This line was removed as it is not used anymore.

Commit:
[TASK] Add PHP Doc Comment for throwing an exception in PollController::publishAction

Commit:
[BUGFIX] Hide pagination when there is only one page to be displayed

Riiiad committed:
[TASK] Update runTests.sh configuration

Riiiad committed:
[TASK] Add gihub actions

Riiiad committed:
[TASK] Steamline configuration and scripts for linters

Riiiad committed:
[TASK] Fix Errors found by checkExceptionCodes command

Riiiad committed:
[TASK] Correct return types, properties and parameters

Resolve PHPStan reported issues till level 5 as preperation for future upgrades

Riiiad committed:
[TASK] Generate baseline for phpstan errors

calien666 commented on Documentation/Settings.cfg:
As the new guides.xml for php rendering is already added, this file can be dropped

Riiiad replied:
Documentation will be ported to new rendering with the Next MR

calien666 commented on Makefile:
Does the rendering of `Build/Scripts/runTests.sh -s renderDocumentation` not work?

Riiiad replied:
Documentation will be ported to new rendering with the Next MR. As the configuration is not working right now

calien666 commented on composer.json:
```diff
@@ -13,50 +13,59 @@
     "license": "GPL-2.0-or-later",
     "require": {
-        "php": "^7.2",
+        "php": "^8.1",
```
Is a minimum of PHP 8.1 really required? According to core support, we should begin with PHP 7.4 up to 8.3. Or is something inside the code preventing PHP 7.4?

calien666 commented on ext_emconf.php:
```diff
@@ -11,14 +11,14 @@
     'constraints' => [
         'depends' => [
-            'typo3' => '9.5.0-10.4.99',
+            'typo3' => '10.4.0-11.5.99',
```
composer.json has requirement only for TYPO3 v11, I guess, we should do the same here.

Additionally, dependencies to other extensions are missing, (fluid_styled_content, numbered_pagination). For legacy installations respecting the dependencies, these should be added here, too.

calien666 commented on rector.php:
```diff
@@ -0,0 +1,58 @@
+    ]);
+
+    $rectorConfig->skip([
+        __DIR__ . '/.Build/*',
```
We should add .ddev here, too, as php scripts could be added inside ddev

calien666 reviewed (commented):
@Riiiad Made some comments, additionally we should add a milestone for v12/v13, adding things like removing renderStatic from ViewHelpers, as this is deprecated since fluid v4.

Riiiad commented on Classes/Domain/Model/BasePoll.php:
```diff
@@ -685,11 +687,11 @@ public function getStatus(): PollStatus
     public function isSchedulePoll(): bool
     {
-        return false !== stripos(get_class($this), 'schedule');
+        return stripos(static::class, 'schedule') !== false;
```
Added todo to refactor with next version

Riiiad commented on Classes/Controller/PollController.php:
```diff
@@ -751,18 +743,14 @@ public function updateAction(\FGTCLB\T3oodle\Domain\Model\BasePoll $poll): void
      * @TYPO3\CMS\Extbase\Annotation\Validate("FGTCLB\T3oodle\Domain\Validator\CustomPollValidator", param="poll")
      */
-    public function deleteAction(\FGTCLB\T3oodle\Domain\Model\BasePoll $poll): void
+    public function deleteAction(BasePoll $poll): void
```
For now yes, will add todo to refactor this with the next release

Riiiad commented on Classes/Controller/PollController.php:
```diff
@@ -272,27 +325,21 @@ public function deleteOwnVoteAction(\FGTCLB\T3oodle\Domain\Model\Vote $vote): vo
      * @\TYPO3\CMS\Extbase\Annotation\IgnoreValidation("poll")
      */
-    public function finishAction(\FGTCLB\T3oodle\Domain\Model\BasePoll $poll, int $option = 0): void
+    public function finishAction(BasePoll $poll, int $option = 0): void
```
Will add a todo this can be done with the controller refactoring in 12 LTS update

Riiiad commented on Classes/Controller/PollController.php:
```diff
@@ -437,24 +472,20 @@ public function editSuggestionAction(
     /**
      * @\TYPO3\CMS\Extbase\Annotation\Validate("FGTCLB\T3oodle\Domain\Validator\SuggestionDtoValidator", param="suggestionDto")
      */
     public function updateSuggestionAction(
```
added todos

Riiiad commented on Classes/Controller/PollController.php:
```diff
@@ -613,60 +632,49 @@ public function createAction(
+     * @throws \FGTCLB\T3oodle\Domain\Permission\AccessDeniedException
      */
-    public function publishAction(\FGTCLB\T3oodle\Domain\Model\BasePoll $poll): void
+    public function publishAction(BasePoll $poll): void
```
Added @todo for next MR

Riiiad commented on Classes/EventListener/UpdatePollSlugListener.php:
```diff
@@ -1,42 +1,63 @@
-class UpdatePollSlug
+class UpdatePollSlugListener
 {
+    protected PollRepository $pollRepository;
+
+    public function injectPollRepository(PollRepository $pollRepository): void
+    {
+        $this->pollRepository = $pollRepository;
+    }
+
+    protected PersistenceManager $persistenceManager;
+
+    public function __construct()
+    {
+        $this->persistenceManager = GeneralUtility::makeInstance(PersistenceManager::class);
+    }
```
Would leave this also for refactoring of the controller.

Riiiad committed:
[TASK] Add todos for refactoring of PollConnector and Resolve version conflicts

Riiiad commented on Classes/Event/ShowPollEvent.php:
```diff
@@ -0,0 +1,77 @@
+
+final class ShowPollEvent
+{
+    private mixed $poll;
```
We have problems when using this and extending the PollController. Will do some testing and add a todo for next release if its not fixed on this branch.

calien666 committed:
[TASK] Update runTests.sh

The extension used a very old runTests script. Copied from
`web-vision/deepltranslate-core` and adopted to the needs of the
extension.

calien666 committed:
[TASK] ext_emconf.php

With rebuilding and refactoring, two dependencies were added in
`composer.json`. In addition to this, the respective dependencies were
added in `ext_emconf.php` ensuriung correct behaviour on legacy
installations.

calien666 committed:
[TASK] Fix TypoScript issues and adopt .editorconfig

TypoScriuptt Linter issued code smells, which were fixed. In addition,
the `.editorconfig` is adopted matching the linter requirements

calien666 committed:
[TASK] Cleanups

Removed superfluous code and installed TYPO3/testingFramework

Used commands:
```shell
composer remove --dev ergebnis/composer-normalize
composer remove --dev ssch/typo3-rector
composer require --dev typo3/testing-framework:^7.1
```

calien666 committed:
[TASK] Move Icons to Registry file

see Changelog:
https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/11.4/Feature-94692-RegisteringIconsViaServiceContainer.html

calien666 committed:
[TASK] Adopt README to markdown

calien666 committed:
[TASK] Streamline phpunit test setup

In addition to the introduction ofthe TYPO3 testing framework, adopt the
configurations matching phpunit 10.5.

calien666 committed:
[!!!][TASK] Remove unused classes

With the goal of getting the Extension working with TYPO3 11 and above
and dropping the support for older TYPO3 versions, remove classes taking
care of older versions, as they are no longer needed

calien666 committed:
[TASK] Deprecate Update wizards

The Update wizards are no longer used and could be removed. As there is
the possibility, someone wants to upgrade from versions below 0.9,
deprecate the classes and mark them to be removed in a later version

calien666 committed:
[TASK] Add basic functional and unit tests

In efforts to a better code maintainability, add small tests for some
parts, testing the functionality of the testing infrastructure and
allow the pipelines run the tests.

calien666 committed:
[TASK] Enable Test runs in CI

calien666 committed:
[TASK] Remove useless composer-for-core-version.sh

calien666 committed:
[TASK] Adjust poll subclass registration

According to old TypoScript extbase configuration, add the class names
as array-keys for usage in TCA.

This allows updating without any breaking inside the database models,
as the models itself are registered with their respective class names
as `recordType`.

calien666 committed:
[TASK] Ensure correct return values

Within v11 it is recommended to use `ResponseInterface` returns.
During the refactoring, most was done, but it was overseen, that the
created responses are used.

calien666 committed:
[TASK] Refactor Event handling

Add event changeable properties and streamline the response.

calien666 committed:
[!!!][TASK] Refactor Permission Event handling

The old slot-adapted solution had a high potential for attacking itself
and thus getting stuck in an event caller loop.

The modification made here places the event handling on the
PollController and allows for finer adjustment, as each step is now
stored individually as an event. On the one hand, this facilitates
troubleshooting and, on the other hand, prevents an unintentional
event-in-event loop, which was previously the case.

At the same time, exception handling is optimised and case-specific
exceptions are thrown, which clearly identify the permissions as the
source of the error and which permission is actually missing at that
point.

The controller itself has been modified so that all actions now return a
ResponseInterface, which will suffice for a later upgrade to TYPO3 v12.

calien666 committed:
[!!!][TASK] Refactor PermissionViewHelper to AbstractConditionViewHelper

As the ViewHelper used deprecated `CompileWithRenderStatic`, refactor to
`AbstractConditionViewHelper` and rework the respective templates.

calien666 committed:
[TASK] Make response changeable by UpdateAfterEvent

Before refactoring to EventListeners, the Slots could create redirects
after manipulating the data.
This technique could be used again, but for getting a clearer state, it
is better to use response manipulation through the event and the
respective Listeners ensuring the response being returned directly
inside the controller instead of some other places.

Create the default redirect response as variable inside the controller
action and override with the manipulated one from the event.

This ensures a clean state and proper possibilities for the manipulation

calien666 committed:
[TASK] Introduce NewPollAllowedEvent

In favour of an early check and exit, add an event to the newAction,
which allows a check for event listeners and an earlier return and
exception throwing than on create action.
